### PR TITLE
UCP/WORKER: Fix removing element pointing to KA::iter and KA::iter_end

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Fix removing element pointing to KA::iter and KA::iter_end

## Why ?

If the iterator and the stop element point to the removed endpoint, the iterator and the stop element should be moved to the previous element and KA round should be completed only if KA is in progress.

## How ?

1. Improve "KA remove EP"-related comments.
2. Reorganize "KA remove EP" code as follows:
```
if ((iter == ep) && (iter_end == ep)) {
    iter = ep->prev;
    iter_end = ep->prev;

    if (lane_map !=0) {
        complete();
    }
} if (iter == ep) {
    lane_map = 0;
    get_next_ep();
} if (iter_end == ep) {
    iter_end = ep->prev;
}
```

### CODE REVIEW

**Comment by dmitrygx** [comment]
@brminich @yosefe could you review pls?

**Comment by yosefe** [comment]
@dmitrygx so the issue was if the current and stop were the same, we skipped over the stop to do another round, while we should have stopped the iteration?

**Comment by dmitrygx** [comment]
> @dmitrygx so the issue was if the current and stop were the same, we skipped over the stop to do another round, while we should have stopped the iteration?

Previously, `round_inprogress` should be `(iter != iter_end) || (lane_map != 0)`, i.e. wasn't handle case when KA iterator on the step element, but it is in-progress.
I reorganized the code to check for `(iter == iter_end == ep)` and move iterators to the previous element and complete KA round if `lane_map != 0` (which means that it is in-progress), otherwise - don't complete, just move iterators to the previous element.

**Comment by yosefe** [comment]
if we just changed the round_inprogress to also check `lane_map != 0`, would it be enough to fix the issue?

**Comment by dmitrygx** [comment]
> if we just changed the round_inprogress to also check `lane_map != 0`, would it be enough to fix the issue?

there is an issue with it.
```
                      iter  iter_end
                      \/   \/
[ EP_1 ] -> [EP_2] -> [EP_3]
```
after removing as it was implemented in master branch:
```
iter_end       iter 
   \/          \/
[ EP_1 ] -> [EP_2]
```
so, they `iter` and `iter_end` are not equal and we will do KA for `EP_2`, but it was already used in the current KA round.

File: src/ucp/core/ucp_worker.c [hunk 1]
```diff
@@ -3090,38 +3089,48 @@ void ucp_worker_keepalive_remove_ep(ucp_ep_h ep)
         return;
     }
 
-    round_inprogress = worker->keepalive.iter != worker->keepalive.iter_end;
+    if ((worker->keepalive.iter == &ucp_ep_ext_gen(ep)->ep_list) &&
+        (worker->keepalive.iter_end == &ucp_ep_ext_gen(ep)->ep_list)) {
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:3093]: `/* we are removing the current and last ep in the keepalive round */`

File: src/ucp/core/ucp_worker.c [hunk 2]
```diff
@@ -3090,38 +3089,48 @@ void ucp_worker_keepalive_remove_ep(ucp_ep_h ep)
...
-    if (worker->keepalive.iter == &ucp_ep_ext_gen(ep)->ep_list) {
-        /* Set lane_map=0 to make sure the endpoint won't be selected again */
-        ucs_debug("worker %p: removed keepalive current ep %p, moving to next",
+        worker->keepalive.iter     = worker->keepalive.iter_end->prev;
+        worker->keepalive.iter_end = worker->keepalive.iter;
+        ucs_assert(worker->keepalive.iter != &ucp_ep_ext_gen(ep)->ep_list);
+
+        if (worker->keepalive.lane_map != 0) {
+            ucs_assert((ucs_get_time() - worker->keepalive.last_round) >=
+                               worker->context->config.ext.keepalive_interval);
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:3106]: ucs_assertv

File: src/ucp/core/ucp_worker.c [hunk 3]
```diff
@@ -3090,38 +3089,48 @@ void ucp_worker_keepalive_remove_ep(ucp_ep_h ep)
...
+                  worker, ep);
 
-    if (worker->keepalive.iter == &ucp_ep_ext_gen(ep)->ep_list) {
-        /* Set lane_map=0 to make sure the endpoint won't be selected again */
-        ucs_debug("worker %p: removed keepalive current ep %p, moving to next",
+        worker->keepalive.iter     = worker->keepalive.iter_end->prev;
+        worker->keepalive.iter_end = worker->keepalive.iter;
+        ucs_assert(worker->keepalive.iter != &ucp_ep_ext_gen(ep)->ep_list);
+
+        if (worker->keepalive.lane_map != 0) {
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:3104]: `/* finish the current round */`

File: src/ucp/core/ucp_worker.c [hunk 4]
```diff
@@ -3090,38 +3089,48 @@ void ucp_worker_keepalive_remove_ep(ucp_ep_h ep)
...
+            ucs_assert((ucs_get_time() - worker->keepalive.last_round) >=
+                               worker->context->config.ext.keepalive_interval);
+            worker->keepalive.lane_map = 0;
+            ucp_worker_keepalive_complete(worker, ucs_get_time());
+        } else {
+            /* Keepalive is not in-progress */
+        }
+    } else if (worker->keepalive.iter == &ucp_ep_ext_gen(ep)->ep_list) {
+        /* Move the iterator to the next element to continue doing keeaplive
+         * starting from the next element during progressing since its lane
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:3114]: /* Removing current ep but not last - move to next ep */

File: src/ucp/core/ucp_worker.c [hunk 5]
```diff
@@ -3090,38 +3089,48 @@ void ucp_worker_keepalive_remove_ep(ucp_ep_h ep)
...
+         */
+        ucs_debug("worker %p: removed the keepalive current endpoint %p",
                   worker, ep);
+
+        /* Set lane_map=0 to make sure the endpoint won't be selected again */
         worker->keepalive.lane_map = 0;
         ucp_worker_keepalive_next_ep(worker);
-        ucs_assert(worker->keepalive.iter != &ucp_ep_ext_gen(ep)->ep_list);
-    }
+    } else if (worker->keepalive.iter_end == &ucp_ep_ext_gen(ep)->ep_list) {
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:3124]: /* Removing last but not current - move iter_end to previous ep */

File: src/ucp/core/ucp_worker.c (hunk 1 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:3094]: done

File: src/ucp/core/ucp_worker.c (hunk 2 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c]: done

File: src/ucp/core/ucp_worker.c [hunk 6]
```diff
@@ -3090,38 +3089,48 @@ void ucp_worker_keepalive_remove_ep(ucp_ep_h ep)
...
+        worker->keepalive.iter_end = worker->keepalive.iter;
+        ucs_assert(worker->keepalive.iter != &ucp_ep_ext_gen(ep)->ep_list);
+
+        if (worker->keepalive.lane_map != 0) {
+            ucs_assert((ucs_get_time() - worker->keepalive.last_round) >=
+                               worker->context->config.ext.keepalive_interval);
+            worker->keepalive.lane_map = 0;
+            ucp_worker_keepalive_complete(worker, ucs_get_time());
+        } else {
+            /* Keepalive is not in-progress */
```
[the remark this answers is not in the document]
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:3121]: done

File: src/ucp/core/ucp_worker.c (hunk 4 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c]: done

File: src/ucp/core/ucp_worker.c (hunk 5 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c:3135]: done

File: src/ucp/core/ucp_worker.c [hunk 7]
```diff
@@ -3091,8 +3092,7 @@ void ucp_worker_keepalive_remove_ep(ucp_ep_h ep)
 
     if ((worker->keepalive.iter == &ucp_ep_ext_gen(ep)->ep_list) &&
         (worker->keepalive.iter_end == &ucp_ep_ext_gen(ep)->ep_list)) {
-        /* Move the iterator and the stop element to the previous element and
-         * complete the current keepalive round if it is in-progress.
+        /* We are removing the current and last endpoint in the keepalive round
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c:3118]: can be a single-line comment?

File: src/ucp/core/ucp_worker.c [hunk 8]
```diff
@@ -3102,29 +3102,39 @@ void ucp_worker_keepalive_remove_ep(ucp_ep_h ep)
         ucs_assert(worker->keepalive.iter != &ucp_ep_ext_gen(ep)->ep_list);
 
         if (worker->keepalive.lane_map != 0) {
-            ucs_assert((ucs_get_time() - worker->keepalive.last_round) >=
-                               worker->context->config.ext.keepalive_interval);
+            /* Finish the current keepalive round */
+#if UCS_ENABLE_ASSERT
+            now = ucs_get_time();
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: pls move the comment from line 3105 to line 3118, and add a one-line comment what this assert is checking

Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c]: I've moved the assertion to separate function to use it in other branches too

File: src/ucp/core/ucp_worker.c (hunk 7 above)
Comment (dmitrygx) [reply src/ucp/core/ucp_worker.c]: done

**Comment by dmitrygx** [comment]
@yosefe the 4th commit  fixes another issue with KA:
```
            iter    iter_end
             |         |
             \/        \/
[EP_1] -> [EP_2] -> [all_eps] -> [EP_3]
```
after removing `EP_2`, it will be:
```
                   iter iter_end
                    |      |
                    \/     \/
[EP_1] -> [EP_2] -> [all_eps] -> [EP_3]
```
so, the next keepalive iteration will move the `iter` to `EP_3` (since `lane_map` is `0` for `all_eps`) which may be already visited.

**Pull request closed without merging**